src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -1207,28 +1213,126 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
+	  ldpp_dout(dpp, 0) << "Failed to delete head object in block directory for: " << obj.objName << ", ret=" << ret << dendl;
+	  return next->delete_obj(dpp, y, flags);
+        }
+      } else {
+	ldpp_dout(dpp, 0) << "Failed to delete head policy entry for: " << obj.objName << ", ret=" << ret << dendl;
+	return next->delete_obj(dpp, y, flags);
+      }
+    } else {
+      ldpp_dout(dpp, 0) << "Failed to delete head object for: " << obj.objName << ", ret=" << ret << dendl;
+      return next->delete_obj(dpp, y, flags);
+    }
 
-  if ((ret = source->driver->get_cache_driver()->del(dpp, source->get_key().get_oid(), y)) < 0) 
-    ldpp_dout(dpp, 0) << "D4NFilterObject::" << __func__ << "(): CacheDriver del() method failed, ret=" << ret << dendl;
+    if ((ret = source->driver->get_obj_dir()->del(dpp, &obj, y)) < 0) {
+      ldpp_dout(dpp, 0) << "Failed to delete object directory entry for: " << obj.objName << ", ret=" << ret << dendl;
+      return next->delete_obj(dpp, y, flags);
+    }
+    
+    block.cacheObj.objName = source->get_key().get_oid();
+    off_t lst = source->get_obj_size();
+    off_t fst = 0;
+
+    do {
+      prefix = source->get_bucket()->get_name() + "_" + version + "_" + obj.objName;
+      if (fst >= lst) {
+	break;
+      }
+
+      off_t cur_size = std::min<off_t>(fst + dpp->get_cct()->_conf->rgw_max_chunk_size, lst);
+      off_t cur_len = cur_size - fst;
+      block.blockID = static_cast<uint64_t>(fst);
+      block.size = static_cast<uint64_t>(cur_len);
+
+      if ((ret = blockDir->get(dpp, &block, y)) < 0) {
+	ldpp_dout(dpp, 0) << "Failed to retrieve directory entry for: " << obj.objName << " blockid: " << fst << " block size: " << cur_len << ", ret=" << ret << dendl;
+	return next->delete_obj(dpp, y, flags);
+      }
+
+      // check if block has been evicted
+      if (std::find(block.hostsList.begin(), block.hostsList.end(), dpp->get_cct()->_conf->rgw_local_cache_address) == block.hostsList.end() ) {
+	ldpp_dout(dpp, 10) << "D4NFilterObject::D4NFilterDeleteOp::" << __func__ << "(): skipping block " << obj.objName << " blockid: " << fst << " block size: " << cur_len << dendl;
+        continue;
+      }
 
-  return next->delete_obj(dpp, y, flags);
+      if (block.dirty == true)
+	prefix = "D_" + prefix;
+
+      std::string oid_in_cache = prefix + "_" + std::to_string(fst) + "_" + std::to_string(cur_len);
+
+      if ((ret = source->driver->get_cache_driver()->delete_data(dpp, oid_in_cache, y)) == 0) { // Sam: do we want del or delete_data here? 
+	if ((ret = source->driver->get_policy_driver()->get_cache_policy()->erase(dpp, policy_prefix + "_" + std::to_string(fst) + "_" + std::to_string(cur_len), y))) {
+	  if ((ret = blockDir->del(dpp, &block, y)) < 0) { 
+	    ldpp_dout(dpp, 0) << "Failed to delete directory entry for: " << obj.objName << " blockid: " << fst << " block size: " << cur_len << ", ret=" << ret << dendl;
+	    return next->delete_obj(dpp, y, flags);
```

Comment:
Since the loop must continue for other blocks, this has to be removed : `next->delete_obj`

Reply:
If the deletion of one block failed, shouldn't the entire operation fail? If it continues as normal and other blocks are properly deleted, the block that failed to be deleted is still taking up space. So instead of `next->delete_obj`, shouldn't I return an error?

Reply:
Even if delete for a block fails and is taking up space, it will eventually be evicted. ok, so you can add a special check for -ENOENT and continue with the loop if this happens and return an error otherwise. Even though the block will eventually be evicted, the error must be communicated back to the end user. So instead of calling next->delete_obj, return the error.